lib/useConfetti.jsx

```diff
@@ -1,188 +1,172 @@
-            tilt: Math.sin(p.tiltAngle) * tiltAmount,
-          }
-        })
+  const onCanvasResize = useCallback(() => {
+    if (canvasRef.current) {
+      // Get canvas bounding box
+      const canvasBBox = canvasRef.current.getBoundingClientRect()
+      const { width, height } = canvasBBox
+      canvasBBoxRef.current = canvasBBox
+
+      // Set canvas size
+      const scale = window.devicePixelRatio
+      canvasRef.current.width = Math.floor(width * scale)
+      canvasRef.current.height = Math.floor(height * scale)
+
+      transformMatrix.current = new DOMMatrix().scale(scale)
+    }
+  }, [])
+
+  // Observe for canvas size changes
+  useEffect(() => {
+    if (canvasRef.current) {
+      // Get canvas context
+      ctxRef.current = canvasRef.current.getContext('2d')
+
+      // Observe for canvas size changes
+      const resizeObserver = new ResizeObserver(onCanvasResize)
+      resizeObserver.observe(canvasRef.current)
+
+      return () => resizeObserver.disconnect()
+    }
+  }, [])
+
+  // Render onto canvas
+  useAnimationFrame(currentTime => {
+    lastFrameTime.current = currentTime
+
+    if (ctxRef.current && canvasBBoxRef.current) {
+      const { width, height } = canvasRef.current
+
+      // Clear canvas
+      ctxRef.current.clearRect(0, 0, width, height)
+
+      // Update particles
+      particles.current = particles.current.reduce((all, p) => {
+        // Still delayed?
+        if (currentTime < p.delayUntil) return [...all, p]
 
         // Remove off-screen particles
-        particlesRef.current = particlesRef.current.filter(p => {
-          if (currentTime < p.delayUntil) return true
-          if (p.x < -killDistance || p.x > width + killDistance) return false
-          if (p.y < -killDistance || p.y > height + killDistance) return false
-          return true
-        })
-
-        // Render particles
-        ctx.clearRect(0, 0, width * scale, height * scale)
-        particlesRef.current.forEach(p => {
-          // Still waiting for delay?
-          if (currentTime < p.delayUntil) return
-
-          const [x, y] = [p.x - canvasLeft, p.y - canvasTop] //[p.x * width, p.y * height]
-          const drawFn = shapeDrawingFunctions[p.shape]
-          drawFn({ p, x, y, ctx, scale })
-        })
-      }
-    }, [])
-
-    return (
-      <canvas
-        ref={canvasRef}
-        style={{
-          display: 'block',
-          position: 'absolute',
-          pointerEvents: 'none',
-          inset: 0,
-          width: '100%',
-          height: '100%',
-        }}
-      />
-    )
+        if (p.x < -killDistance/window.devicePixelRatio
+          || p.x > (width + killDistance)/window.devicePixelRatio
+          || p.y < -killDistance/window.devicePixelRatio
+          || p.y > (height + killDistance)/window.devicePixelRatio
+        ) return all
+
+        // Render particle
+        const [x, y] = [p.x - canvasBBoxRef.current.left, p.y - canvasBBoxRef.current.top]
```

Comment:
I think using `const ctx =` helps with readability and also can be used to do the shorthand object syntax.